# Assignment  Practice Text classification with Naive Bayes 

## Notebook made by   (If not filled in correctly: 0 pts for assignment)

__Name(s)__: 

__Student id(s)__ : 

### Pledge (taken from [Coursera's Honor Code](https://www.coursera.org/about/terms/honorcode) )



Put here a selfie with your photo where you hold a signed paper with the following text: (if this is team work, put two selfies here). The link must be to some place on the web, not to a local file. **Assignments without the selfies will not be graded and receive 0 points.**

> My answers to homework, quizzes and exams will be my own work (except for assignments that explicitly permit collaboration).

>I will not make solutions to homework, quizzes or exams available to anyone else. This includes both solutions written by me, as well as any official solutions provided by the course staff.

>I will not engage in any other activities that will dishonestly improve my results or dishonestly improve/hurt the results of others.

<img src='link to your selfie'/>

# Some example code

In [2]:
import pandas as pd

# normal KVR.csv.gz
# subset = KVR1000.csv.gz
kvrdf= pd.read_csv('KVR.csv.gz', 
                   sep='\t', encoding='utf-8',
                   index_col=0, 
                   names=['jaar', 'partij','titel','vraag','antwoord','ministerie']) 
print (kvrdf.shape)
kvrdf.head()

(40516, 6)


jaar partij  \
KVR1000.xml     1994   PvdA   
 KVR10000.xml   1999   PvdA   
 KVR10001.xml   1999     SP   
 KVR10002.xml   1999   PvdA   
 KVR10003.xml   1999   PvdA   

                                                           titel  \
KVR1000.xml     De vragen betreffen de betrouwbaarheid van de...   
 KVR10000.xml   Vragen naar aanleiding van berichten (uitzend...   
 KVR10001.xml   Vragen naar aanleiding van de berichten "Nede...   
 KVR10002.xml   Vragen over de gebrekkige opvang van verpleeg...   
 KVR10003.xml   Vragen over onbetrouwbaarheid van filemeldingen.   

                                                           vraag  \
KVR1000.xml     Hebt u kennisgenomen van het televisieprogram...   
 KVR10000.xml   Kent u de berichten over de situatie in de Me...   
 KVR10001.xml   Kent u de berichten «Nederland steunt de Soeh...   
 KVR10002.xml   Kent u het bericht over onderzoek van Nu91 me...   
 KVR10003.xml   Hebt u kennisgenomen van de berichten over de...   

                                                        antwoord  \
KVR1000.xml     Ja. Het bedoelde geluidmeetpunt is eigendom v...   
 KVR10000.xml                                                      
 KVR10001.xml                                                      
 KVR10002.xml   Ja. Het onderzoek van NU’91 wijst uit dat het...   
 KVR10003.xml   Ja. Nee. Door de waarnemers van het Algemeen ...   

                                       ministerie  
KVR1000.xml                 Verkeer en Waterstaat  
 KVR10000.xml                            Justitie  
 KVR10001.xml                           Financien  
 KVR10002.xml   Volksgezondheid, Welzijn en Sport  
 KVR10003.xml               Verkeer en Waterstaat

In [3]:
len(kvrdf)

40516

In [4]:
# USE FIRST 27000 entries (later entries have no label and are therefore not helpful)

# use 0.8 * 27000 = 21600 entries for training
# use 0.2 * 27000 = 5400 for testing

In [5]:
kvrdf.ministerie.value_counts().head(10)

 Justitie (JUS)                                                   3219
 Volksgezondheid, Welzijn en Sport (VWS)                          2630
 Buitenlandse Zaken (BUZA)                                        1796
 Verkeer en Waterstaat (VW)                                       1441
 Justitie                                                         1333
 Sociale Zaken en Werkgelegenheid (SZW)                           1231
 Onderwijs, Cultuur en Wetenschappen (OCW)                        1187
 Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)     984
 Financiën (FIN)                                                   960
 Volksgezondheid, Welzijn en Sport                                 951
Name: ministerie, dtype: int64

In [6]:
import nltk
import string
from collections import Counter

# Join all titels of the questions together
# use first 21600 entries for training
allvragen = '\n'.join(str(v) for v in list(kvrdf.titel[:21600]))

# remove punctuation
allvragen = allvragen.translate(str.maketrans({key: None for key in string.punctuation}))


print (len(allvragen), allvragen[:1000])

5095344  De vragen betreffen de betrouwbaarheid van de geluidsmetingen op Schiphol
 Vragen naar aanleiding van berichten uitzending NOVA 17 juli 1999 over onderzoeken van de Rijksrecherche waaruit blijt dat er twijfels bestaan over het professioneel gedrag van vier medewerkers van de kliniek
 Vragen naar aanleiding van de berichten Nederland steunt Soehartos fiscaal en Wij hebben een naam hoog te houden in NRC van 22 juli 1999
 Vragen over de gebrekkige opvang van verpleegkundigen met traumatische ervaringen door ziekenhuizen en zorginstellingen
 Vragen over onbetrouwbaarheid van filemeldingen
 Vragen naar aanleiding van het bericht in de Volkskrant van 7 juli 1999 dat het RIVM onjuiste gegevens zou hebben gehanteerd en gepresenteerd wat betreft de milieuaspecten van de Betuwelijn
 
 Vragen naar aanleiding van het artikel Eilander artsen vechten voor waarneming in de Leeuwarder Courant van 26 juni 1999 Deze vragen dienen ter aanvulling op eerdere vragen van de leden Van BlerckWoerdman 

In [7]:
# Tokenize and turn into a bag of words
BoW = Counter(nltk.word_tokenize(allvragen))
# show the top 20 most used words 
BoW.most_common(20)

[('van', 53325),
 ('de', 50554),
 ('het', 24522),
 ('in', 18867),
 ('en', 15917),
 ('over', 15288),
 ('Vragen', 15243),
 ('een', 14603),
 ('naar', 11928),
 ('op', 9832),
 ('aanleiding', 9038),
 ('dat', 8645),
 ('voor', 7654),
 ('te', 7440),
 ('met', 5642),
 ('vragen', 5351),
 ('is', 4742),
 ('aan', 4642),
 ('die', 4612),
 ('om', 4299)]

In [8]:
from nltk.corpus import stopwords

DutchStop= stopwords.words('dutch')
print (len(DutchStop))

BoW = Counter([w for w in nltk.word_tokenize(allvragen) if not w in set(DutchStop)])
# show the top 20 most used words 
BoW.most_common(20)

101


[('Vragen', 15243),
 ('aanleiding', 9038),
 ('vragen', 5351),
 ('De', 3524),
 ('bericht', 3356),
 ('2007', 2185),
 ('minister', 2129),
 ('artikel', 2128),
 ('eerdere', 2008),
 ('Indiener', 1908),
 ('Nederland', 1835),
 ('Nederlandse', 1833),
 ('ingezonden', 1800),
 ('nr', 1558),
 ('lid', 1374),
 ('2006', 1314),
 ('Handelingen', 1305),
 ('Dagblad', 1262),
 ('2005', 1249),
 ('Aanhangsel', 1221)]

# Practice Text classification with Naive Bayes  
        
        
        
<h3>Abstract</h3>
<p>We will do text classification on a collection of Dutch parliamentary questions.
    The website <a href="https://zoek.officielebekendmakingen.nl/zoeken/parlementaire_documenten">officielebekendmakingen.nl</a>lets you search in "kamervragen".
    You can donwload
    <a href='http://data.politicalmashup.nl/kamervragen/PoliDocs_Kamervragen.zip'>this zipfile with Kamervragen in XML</a>
    to see some of the  data in XML format. 
    It also contains style sheets to show the XML well in a browser.  
    The <a href='http://maartenmarx.nl/teaching/zoekmachines/LectureNotes/MySQL/'>MYSQL directory</a> contains an <a href='http://maartenmarx.nl/teaching/zoekmachines/LectureNotes/MySQL/KVR14807.xml'>example   Kamervraag XML file</a>. Note that in your browser you see the result of applying stylesheets. So choose View Source or open it in an editor.</p>

<h3>First exploration</h3>

See below.

<h2>Exercise</h2>

<p>We will use the fields in elements of the form <tt> &lt;item attribuut="Afkomstig_van"></tt> as our classes. 
    These are the ministeries to whom the question is addressed.
    An example is 
    <pre>
        &lt;item attribuut="Afkomstig_van">Landbouw, Natuurbeheer en Visserij (LNV)&lt;/item>
    </pre>
    Note that these labels are <strong>not normalized</strong>, see e.g. the counts below:
    <pre>
Justitie (JUS)                                                   3219
Volksgezondheid, Welzijn en Sport (VWS)                          2630
Buitenlandse Zaken (BUZA)                                        1796
Verkeer en Waterstaat (VW)                                       1441
Justitie                                                         1333
Sociale Zaken en Werkgelegenheid (SZW)                           1231
Onderwijs, Cultuur en Wetenschappen (OCW)                        1187
Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)     984
Financiën (FIN)                                                   960
Volksgezondheid, Welzijn en Sport                                 951
Economische Zaken (EZ)                                            946
Buitenlandse Zaken                                                753
Binnenlandse Zaken en Koninkrijksrelaties (BZK)                   725
Verkeer en Waterstaat                                             724
Defensie (DEF)                                                    646
Sociale Zaken en Werkgelegenheid                                  607
Landbouw, Natuurbeheer en Visserij (LNV)                          586
Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer            554
Onderwijs, Cultuur en Wetenschappen                               532
Vreemdelingenzaken en Integratie (VI)                             466
    </pre>
</p>

  <ol>
      <li>Normalize the values for "ministerie" and choose 10 ministeries to work with. </li>
      <li>Implement the two algorithms in Fig MRS.13.2, using your earlier code for creating term and document frequencies.
      It might be easier to use the representation and formula given in MRS section 13.4.1.</li>
      <li>On this collection, train NB text classifiers for 10 different classes with enough and interesting data.</li>
      <li>Compute for each term and each of your 10 classes its utility for that class using mutual information.</li>
      <li>For each class, show the top 10 words as in Figure 13.7 in MRS.</li>
      <li>Evaluate your classifiers using Precision, Recall and F1. (
           <br/>
          Give a table in which you show these values for using the top 10, top 100 terms and all terms, for all of your 10 classes.
          Thus do feature selection per class, and use for each class the top n best features for that class. 
          <br/>
      Also show the microaverage(s) for all 10 classes together.
      <br/>
      If you like you can also present this in a figure like MRS.13.8. 
      Then compute the F1 measure for the same number of terms as in that figure.</li>
      <li>Reflect and report briefly about your choices in this process and about the obtained results. </li>
  </ol>

<h3>Training/Testing</h3>
<p>It is important that you do not test your classifier using documents that have also been used in training.
    So split up your collection in a training set and a test set. A 80%-20% split is reasonable.

<br/>
    If you have too little data you can use 5 or <a href="http://en.wikipedia.org/wiki/Cross-validation_(statistics)#k-fold_cross-validation">10-fold cross validation</a>.</p>

<h2>Form of presentation</h2>
<ul>
    <li>Make slides or wikipages and have your system running and be able to accept documents from the web. </li>
    <li>Create one or two slides or wikipages for each of the sub exercises listed above.
</li>
<li>Make it clear in the heading of the slides which sub exercises you talk about.</li>
    <li>Show running code with one or two  good examples (a TP of course, but also a FP and an error-analysis is nice to show). </li>

</ul>

<h2>Form of handing in your final product</h2>
<ul>
    <li>An IPython notebook would be perfect, with clear indications which part of the code answers which subquestion.</li>
    <li>A clear git repo, with good comments and a clear separation and indication what code does what is also fine.</li>
    <li> You are free to program in whatever language you prefer.</li>
</ul>


### Question 1

In [9]:
kvrdf.ministerie[:21600].value_counts().head(20)

 Justitie (JUS)                                                   3006
 Volksgezondheid, Welzijn en Sport (VWS)                          2471
 Buitenlandse Zaken (BUZA)                                        1704
 Verkeer en Waterstaat (VW)                                       1330
 Sociale Zaken en Werkgelegenheid (SZW)                           1130
 Onderwijs, Cultuur en Wetenschappen (OCW)                        1113
 Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)     902
 Financiën (FIN)                                                   891
 Economische Zaken (EZ)                                            877
 Binnenlandse Zaken en Koninkrijksrelaties (BZK)                   716
 Defensie (DEF)                                                    590
 Justitie                                                          517
 Landbouw, Natuurbeheer en Visserij (LNV)                          502
 Vreemdelingenzaken en Integratie (VI)                             466
 Landb

In [10]:
normalized_ministerie = kvrdf.ministerie[:21600].value_counts().head(10) / sum(kvrdf.ministerie[:21600].value_counts())

print(normalized_ministerie)
# choice of ministeries: top 10

 Justitie (JUS)                                                   0.139742
 Volksgezondheid, Welzijn en Sport (VWS)                          0.114871
 Buitenlandse Zaken (BUZA)                                        0.079215
 Verkeer en Waterstaat (VW)                                       0.061829
 Sociale Zaken en Werkgelegenheid (SZW)                           0.052531
 Onderwijs, Cultuur en Wetenschappen (OCW)                        0.051741
 Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)    0.041932
 Financiën (FIN)                                                  0.041421
 Economische Zaken (EZ)                                           0.040770
 Binnenlandse Zaken en Koninkrijksrelaties (BZK)                  0.033285
Name: ministerie, dtype: float64


In [11]:
ministerie_names = list(normalized_ministerie.keys())

### Question 2

In [12]:
# for all classes, concatenate all 'vragen'

V = list(BoW)
text = []
for c in range(len(normalized_ministerie)):
    text.append([])
    prior_c = normalized_ministerie[c]
    ministerie = kvrdf.ministerie[:21600].value_counts().keys()[c]
    print(c+1)
    
    for d in range(len(kvrdf.ministerie[:21600])):
        # only check first 8 characters of string.
        # assumption: if first 8 characters match --> same ministerie 
        # this way, we solve 'Justitie' not equal to 'Justitie (JUS)'
        if(str(ministerie)[:8] == str(kvrdf.ministerie[d])[:8]):
            text_document = kvrdf.titel[d].translate(str.maketrans({key: None for key in string.punctuation}))
            text[c].append(text_document)

1
2
3
4
5
6
7
8
9
10


In [13]:
# make BoW representation
def countTokensOfTerm(text, n_class):
    wordstring = ''.join(text[n_class])
    
    # Tokenize and turn into a bag of words
    DutchStop= stopwords.words('dutch')
    BoW_class = Counter([w for w in nltk.word_tokenize(wordstring) if not w in set(DutchStop)])
   
    return BoW_class

In [14]:
BoW_classes = list()

# for all 10 classes, make BoW representation
for i in range(len(normalized_ministerie)):
    BoW_classes.append(countTokensOfTerm(text, i))

In [15]:
# calculate conditional probability of term t in class c    
def condprob(c, t):
    vocabulary_length = len(BoW)
    
    Tct = BoW_classes[c][t]
    #print("Tct: ",Tct)
    
    # Tct_neg = sum of everything not Tct
    Tct_neg = sum(BoW_classes[c].values()) - Tct
    #print("Tct_neg: ", Tct_neg)
   
    #conprob[c][t] = 
    return (Tct + 1) / (Tct_neg +  vocabulary_length)


print(condprob(0, 'kinderporno'))
print(condprob(1, 'kinderporno'))
print(condprob(2, 'kinderporno'))
print(condprob(3, 'kinderporno'))
print(condprob(4, 'kinderporno'))
print(condprob(5, 'kinderporno'))


# intuitief klopt want Justitie hoge score op 'kinderporno'
    

0.0005246152821264406
9.490097083693167e-06
1.1080209637566344e-05
1.3042060645582003e-05
1.3734565781702812e-05
1.2829394709157622e-05


In [16]:
print(condprob(0, 'politie'))
print(condprob(1, 'politie'))
print(condprob(2, 'politie'))
print(condprob(3, 'politie'))
print(condprob(4, 'politie'))
print(condprob(5, 'politie'))

0.0019346392148033257
0.0001423703717764975
0.00023273597765734614
0.00013043591683405942
1.3734565781702812e-05
8.981267641775726e-05


In [17]:
import math

# USE FIRST 27000 entries (later entries have no 'ministerie' label and are therefore not helpful)
# use 0.8 * 27000 = 21600 entries for training
# use 0.2 * 27000 = 5400 for testing


# input: d (to be classified document), classes_prior (classes and their probabilities)
# output: scores for all d - class pairs
def applyMultinomialNB(d, classes_prior):
    # get tokens out of to_be_classified
    BoW_d = Counter([w for w in nltk.word_tokenize(str(d).translate(str.maketrans({key: None for key in string.punctuation}))) if not w in set(stopwords.words('dutch'))])
    scores = dict()
    
    # scores = p_c + conprob
    for i in range(len(classes_prior)):
        # p_c
        scores[i] = math.log10(classes_prior[i])

        # for every term calculate conprob
        for token in BoW_d:
                # calculate conprob of token and class * n of occurrences of token
                scores[i] += math.log10(condprob(i, token)) * BoW_d[token]
                    
    return scores


testdoc = kvrdf.titel.head(40515)[22201:][0]

scores = applyMultinomialNB(testdoc, normalized_ministerie)

sorted_scores = sorted(scores.items(), key=lambda x : x[1], reverse=True)
for minis, score in sorted_scores:
    print(ministerie_names[minis], ': ',score)

 Verkeer en Waterstaat (VW) :  -35.58159331334153
 Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM) :  -36.370387237895926
 Economische Zaken (EZ) :  -37.693770259335636
 Sociale Zaken en Werkgelegenheid (SZW) :  -37.83026213025033
 Financiën (FIN) :  -37.88609738517333
 Binnenlandse Zaken en Koninkrijksrelaties (BZK) :  -37.94160796938409
 Onderwijs, Cultuur en Wetenschappen (OCW) :  -38.09790104283042
 Justitie (JUS) :  -38.20048894076771
 Volksgezondheid, Welzijn en Sport (VWS) :  -38.38476427411192
 Buitenlandse Zaken (BUZA) :  -38.82533432946341


In [18]:
kvrdf.titel.head(40515)[22201:][0]

' Betreft een gedoogbeschikking voor het lozen van retourperswater, door Rijkswaterstaat afgegeven voor een slibdepot in het Ketelmeer.'

In [19]:
kvrdf.ministerie.head(40515)[22201:][0]

' Verkeer en Waterstaat'

### Question 4

In [20]:
def vraag_in_c(clss):
    in_c = list()

    for i in range(len(kvrdf.titel[:21600])):
        current_ministerie = str(kvrdf.ministerie[i])
        ministerie_clss = ministerie_names[clss] 
        # use first 8 characters of string for comparison
        if current_ministerie[:8] == ministerie_clss[:8]:
            in_c.append(i)
    return in_c
    
# loop through all classes, fill list with all 'vragen' per class
vragen_in_c = list()
for i in range(len(normalized_ministerie)):
    vragen_in_c.append(vraag_in_c(i))

In [21]:
# make index that looks like:
# {term : [list with positions of vragen where term occurs]}

from collections import defaultdict

def index_t():
    index_t = defaultdict(list)
    
    for i in range(len(kvrdf.titel[:21600])):
        vraag = [w for w in nltk.word_tokenize(str(kvrdf.titel[i]).translate(str.maketrans({key: None for key in string.punctuation}))) if not w in set(DutchStop)]
        for t in vraag:
            index_t.setdefault(t,list()).append(i)
    
    return index_t

index_t = index_t()

In [114]:
#Compute for each term and each of your 10 classes its utility for that class using mutual information.
import time

def mutualInformation(t, c):
    N = 21600
    all_questions = set(list(range(N)))
        
    # find all 'vragen' in c
    v_in_c = set(vragen_in_c[c])
    
    # find all 'vragen' not in c
    v_not_in_c = all_questions - v_in_c
    
    # find all 'vragen' that contain term
    v_with_t = index_t[t]
    
    N11 = len(v_in_c.intersection(v_with_t))
    N01 = len(v_in_c) - N11
    N10 = len(v_not_in_c.intersection(v_with_t))
    N00 = len(v_not_in_c) - N10
    
#     Misschien een oplossing dit onderstaande voor math domain error
    if(N11 == 0 or N10 == 0 or N01 == 0 or N00 == 0):
        return 0
    
    
    
    return (N11/N) * math.log2((N*N11) / ((N11+N10)*(N11+N01))) + (N01/N) * math.log2((N*N01) / ((N01+N00)*(N11+N01))) + (N10/N) * math.log2((N*N10) / ((N11+N10)*(N10+N00))) + (N00/N) * math.log2((N*N00) / ((N01+N00)*(N10+N00)))

mutualInformation('vervolgen', 0)


# problem: sometimes math error (divide by 0 etc)

0.0021075840763242278

In [23]:
for i in index_t['vervolgen'][:10]:
    print (kvrdf.ministerie[i])

 Justitie (JUS)
 Justitie (JUS)
 Justitie (JUS)
 Justitie (JUS)
 Justitie (JUS)
 Justitie (JUS)
 Justitie (JUS)
 Justitie (JUS)
 Binnenlandse Zaken en Koninkrijksrelaties (BZK)
 Binnenlandse Zaken en Koninkrijksrelaties (BZK)


### Question 5

In [120]:
#For each class, show the top 10 words as in Figure 13.7 in MRS.

V = list(BoW)
# for c in range(len(normalized_ministerie)):
for c in range(len(normalized_ministerie)):
    scores = dict()
    print("Calculating mutual information...")
    for t in V:
        scores[t] = mutualInformation(t, c)
    top_ten = sorted(scores.items(), key = lambda x: x[1], reverse=True)[:10]
    print(ministerie_names[c])
    print(top_ten)
    print("\n\n")


Calculating mutual information...
 Justitie (JUS)
[('Justitie', 0.009360525334229677), ('OM', 0.008434641662139362), ('Openbaar', 0.0075673146694576355), ('politie', 0.005479477009228581), ('asielzoekers', 0.005303434704004404), ('Jus', 0.004953267028319425), ('justitie', 0.004916605392324226), ('verdachte', 0.004571309332002282), ('Ministerie', 0.004108170588168028), ('criminelen', 0.003990730350736283)]



Calculating mutual information...
 Volksgezondheid, Welzijn en Sport (VWS)
[('zorg', 0.02143501844372219), ('VWS', 0.01744683865257183), ('patiënten', 0.017121431578551186), ('ziekenhuizen', 0.015381366592052248), ('Medisch', 0.010519907090232102), ('zorgverzekeraars', 0.009425743878921497), ('artsen', 0.009006320943273288), ('Gezondheidszorg', 0.00823600273196718), ('huisartsen', 0.007489049464990485), ('medische', 0.006280615122969339)]



Calculating mutual information...
 Buitenlandse Zaken (BUZA)
[('regering', 0.015701832424493585), ('Buitenlandse', 0.009824069741748476), ('Is